xwiki-pro-macros-api/src/main/java/com/xwiki/macros/userlist/internal/macro/UserListMacro.java

```diff
@@ -0,0 +1,91 @@
+/*
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU Lesser General Public License as
+ * published by the Free Software Foundation; either version 2.1 of
+ * the License, or (at your option) any later version.
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package com.xwiki.macros.userlist.internal.macro;
+
+import java.util.List;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+import javax.script.ScriptContext;
+
+import org.xwiki.component.annotation.Component;
+import org.xwiki.rendering.block.Block;
+import org.xwiki.rendering.macro.AbstractMacro;
+import org.xwiki.rendering.macro.MacroExecutionException;
+import org.xwiki.rendering.transformation.MacroTransformationContext;
+import org.xwiki.script.ScriptContextManager;
+import org.xwiki.template.Template;
+import org.xwiki.template.TemplateManager;
+
+import com.xwiki.macros.userlist.macro.UserListMacroParameters;
+
+/**
+ * This macro displays a list of users with their name and avatar.
+ *
+ * @version $Id$
+ */
+
+@Component
+@Named("userList")
+@Singleton
+public class UserListMacro extends AbstractMacro<UserListMacroParameters>
+{
+    @Inject
+    private TemplateManager templateManager;
+
+    @Inject
+    private ScriptContextManager scriptContextManager;
+
+    /**
+     * Create and initialize the descriptor of the macro.
+     */
+    public UserListMacro()
+    {
+        super("User list", "Displays a custom list of users with their avatar",
+            UserListMacroParameters.class);
+    }
+
+    @Override
+    public List<Block> execute(UserListMacroParameters parameters, String content, MacroTransformationContext context)
+        throws MacroExecutionException
+    {
+        Template customTemplate = this.templateManager.getTemplate("html_displayer/userreferencelist/view.vm");
+
+        try {
+            this.bindParameters(parameters);
+            return this.templateManager.execute(customTemplate).getChildren();
+        } catch (Exception e) {
+            throw new MacroExecutionException("Failed to render the userList viewer template.", e);
+        }
```

Comment:
The path to the HTML displayer template, it's name and the way the parameters are bind is an implementation detail of the HTML displayer. Your code shouldn't be aware of that (i.e. you shouldn't have to update this code if the HTML displayer implementation changes). Why not use directly the HTML displayer API? See https://github.com/xwiki/xwiki-platform/blob/master/xwiki-platform-core/xwiki-platform-display/xwiki-platform-display-html/src/main/java/org/xwiki/rendering/display/html/script/HTMLDisplayerScriptService.java .

Reply:
Thanks for the explanation. I modified the code to use the `HTMLDisplayerManager`. Not sure this is completely in line with what you meant though since you mentioned the `HTMLDisplayerScriptService`.  What do you think?

Reply:
Using this approach, I'm running into an issue with the avatar: the HTML for the avatar image is not generated correctly, partially interpreted as wiki syntax it seems, with the following result scheme (with the unexpected `wikigeneratedid` class):

```
<img src="/xwiki/bin/download/Sandbox/UserList/WebHome/%2Fxwiki%2Fbin%2Fskin%2Fresources%2Ficons%2Fxwiki%2Fnoavatar.png%3Fcache-version%3D1654607551492" alt="Bob" title="Bob" class="avatar avatar_50 wikigeneratedid" id="I2Fxwiki2Fbin2Fskin2Fresources2Ficons2Fxwiki2Fnoavatar.png3Fcache-version3D1654607551492">```
```

I could not understand why so far but maybe I'm not using the `HTMLDisplayer*` services along what you had in mind so I'll 
wait for your input about that before digging further.

Reply:
I fixed the avatar issue by generating a `new RawBlock(html, Syntax.HTML_5_0)` directly rather than using an html parser (the initial `@Named("xhtml/5")` Parser was wrong, but even when using a `html/5.0` I was not getting the expected blocks, I don't understand well why).